File: lib/features/results/data/models/trip.dart
import '../../../../core/constants/app_constants.dart';

/// Trip model for search results
class Trip {
  final String id;
  final String carrierId;
  final String carrierName;
  final String? carrierLogo;
  final TransportMode mode;
  final String from;
  final String fromCode;
  final String to;
  final String toCode;
  final DateTime departAt;
  final DateTime arriveAt;
  final Duration duration;
  final double basePrice;
  final String currency;
  final List<String> stops;
  final List<ClassOption> classOptions;
  final Map<String, dynamic> metadata;

  const Trip({
    required this.id,
    required this.carrierId,
    required this.carrierName,
    this.carrierLogo,
    required this.mode,
    required this.from,
    required this.fromCode,
    required this.to,
    required this.toCode,
    required this.departAt,
    required this.arriveAt,
    required this.duration,
    required this.basePrice,
    required this.currency,
    required this.stops,
    required this.classOptions,
    required this.metadata,
  });

  factory Trip.fromJson(Map<String, dynamic> json) {
    return Trip(
      id: json['id'] as String,
      carrierId: json['carrierId'] as String,
      carrierName: json['carrierName'] as String,
      carrierLogo: json['carrierLogo'] as String?,
      mode: TransportMode.values.firstWhere(
        (e) => e.value == json['mode'],
        orElse: () => TransportMode.flight,
      ),
      from: json['from'] as String,
      fromCode: json['fromCode'] as String,
      to: json['to'] as String,
      toCode: json['toCode'] as String,
      departAt: DateTime.parse(json['departAt'] as String),
      arriveAt: DateTime.parse(json['arriveAt'] as String),
      duration: Duration(minutes: json['duration'] as int),
      basePrice: (json['basePrice'] as num).toDouble(),
      currency: json['currency'] as String,
      stops: List<String>.from(json['stops'] as List),
      classOptions: (json['classOptions'] as List)
          .map((e) => ClassOption.fromJson(e as Map<String, dynamic>))
          .toList(),
      metadata: Map<String, dynamic>.from(json['metadata'] as Map),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'carrierId': carrierId,
      'carrierName': carrierName,
      'carrierLogo': carrierLogo,
      'mode': mode.value,
      'from': from,
      'fromCode': fromCode,
      'to': to,
      'toCode': toCode,
      'departAt': departAt.toIso8601String(),
      'arriveAt': arriveAt.toIso8601String(),
      'duration': duration.inMinutes,
      'basePrice': basePrice,
      'currency': currency,
      'stops': stops,
      'classOptions': classOptions.map((e) => e.toJson()).toList(),
      'metadata': metadata,
    };
  }

  Trip copyWith({
    String? id,
    String? carrierId,
    String? carrierName,
    String? carrierLogo,
    TransportMode? mode,
    String? from,
    String? fromCode,
    String? to,
    String? toCode,
    DateTime? departAt,
    DateTime? arriveAt,
    Duration? duration,
    double? basePrice,
    String? currency,
    List<String>? stops,
    List<ClassOption>? classOptions,
    Map<String, dynamic>? metadata,
  }) {
    return Trip(
      id: id ?? this.id,
      carrierId: carrierId ?? this.carrierId,
      carrierName: carrierName ?? this.carrierName,
      carrierLogo: carrierLogo ?? this.carrierLogo,
      mode: mode ?? this.mode,
      from: from ?? this.from,
      fromCode: fromCode ?? this.fromCode,
      to: to ?? this.to,
      toCode: toCode ?? this.toCode,
      departAt: departAt ?? this.departAt,
      arriveAt: arriveAt ?? this.arriveAt,
      duration: duration ?? this.duration,
      basePrice: basePrice ?? this.basePrice,
      currency: currency ?? this.currency,
      stops: stops ?? this.stops,
      classOptions: classOptions ?? this.classOptions,
      metadata: metadata ?? this.metadata,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Trip &&
        other.id == id &&
        other.carrierId == carrierId &&
        other.carrierName == carrierName &&
        other.carrierLogo == carrierLogo &&
        other.mode == mode &&
        other.from == from &&
        other.fromCode == fromCode &&
        other.to == to &&
        other.toCode == toCode &&
        other.departAt == departAt &&
        other.arriveAt == arriveAt &&
        other.duration == duration &&
        other.basePrice == basePrice &&
        other.currency == currency &&
        _listEquals(other.stops, stops) &&
        _listEquals(other.classOptions, classOptions) &&
        _mapEquals(other.metadata, metadata);
  }

  @override
  int get hashCode {
    return Object.hash(
      id,
      carrierId,
      carrierName,
      carrierLogo,
      mode,
      from,
      fromCode,
      to,
      toCode,
      departAt,
      arriveAt,
      duration,
      basePrice,
      currency,
      Object.hashAll(stops),
      Object.hashAll(classOptions),
      Object.hashAll(metadata.entries),
    );
  }

  bool _listEquals<T>(List<T> a, List<T> b) {
    if (a.length != b.length) return false;
    for (int i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }

  bool _mapEquals<K, V>(Map<K, V> a, Map<K, V> b) {
    if (a.length != b.length) return false;
    for (final key in a.keys) {
      if (!b.containsKey(key) || a[key] != b[key]) return false;
    }
    return true;
  }
}

/// Class option for different service classes
class ClassOption {
  final String id;
  final String name;
  final String displayName;
  final double price;
  final double priceAddon;
  final List<String> amenities;
  final Map<String, dynamic> metadata;

  const ClassOption({
    required this.id,
    required this.name,
    required this.displayName,
    required this.price,
    required this.priceAddon,
    required this.amenities,
    required this.metadata,
  });

  factory ClassOption.fromJson(Map<String, dynamic> json) {
    return ClassOption(
      id: json['id'] as String,
      name: json['name'] as String,
      displayName: json['displayName'] as String,
      price: (json['price'] as num).toDouble(),
      priceAddon: (json['priceAddon'] as num).toDouble(),
      amenities: List<String>.from(json['amenities'] as List),
      metadata: Map<String, dynamic>.from(json['metadata'] as Map),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'displayName': displayName,
      'price': price,
      'priceAddon': priceAddon,
      'amenities': amenities,
      'metadata': metadata,
    };
  }

  ClassOption copyWith({
    String? id,
    String? name,
    String? displayName,
    double? price,
    double? priceAddon,
    List<String>? amenities,
    Map<String, dynamic>? metadata,
  }) {
    return ClassOption(
      id: id ?? this.id,
      name: name ?? this.name,
      displayName: displayName ?? this.displayName,
      price: price ?? this.price,
      priceAddon: priceAddon ?? this.priceAddon,
      amenities: amenities ?? this.amenities,
      metadata: metadata ?? this.metadata,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is ClassOption &&
        other.id == id &&
        other.name == name &&
        other.displayName == displayName &&
        other.price == price &&
        other.priceAddon == priceAddon &&
        _listEquals(other.amenities, amenities) &&
        _mapEquals(other.metadata, metadata);
  }

  @override
  int get hashCode {
    return Object.hash(
      id,
      name,
      displayName,
      price,
      priceAddon,
      Object.hashAll(amenities),
      Object.hashAll(metadata.entries),
    );
  }

  bool _listEquals<T>(List<T> a, List<T> b) {
    if (a.length != b.length) return false;
    for (int i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }

  bool _mapEquals<K, V>(Map<K, V> a, Map<K, V> b) {
    if (a.length != b.length) return false;
    for (final key in a.keys) {
      if (!b.containsKey(key) || a[key] != b[key]) return false;
    }
    return true;
  }
}

File: backend/src/models/Trip.js
const mongoose = require('mongoose');

const classOptionSchema = new mongoose.Schema({
  id: {
    type: String,
    required: true
  },
  name: {
    type: String,
    required: true
  },
  description: String,
  price: {
    type: Number,
    required: true,
    min: 0
  },
  currency: {
    type: String,
    required: true,
    default: 'VND'
  },
  availableSeats: {
    type: Number,
    required: true,
    min: 0
  },
  totalSeats: {
    type: Number,
    required: true,
    min: 0
  },
  amenities: [String],
  refundable: {
    type: Boolean,
    default: true
  },
  changeable: {
    type: Boolean,
    default: true
  }
}, { _id: false });

const stopSchema = new mongoose.Schema({
  code: {
    type: String,
    required: true
  },
  name: {
    type: String,
    required: true
  },
  arriveAt: Date,
  departAt: Date,
  duration: {
    type: Number, // Duration in minutes
    default: 0
  }
}, { _id: false });

const tripSchema = new mongoose.Schema({
  carrierId: {
    type: String,
    required: [true, 'ID nhà cung cấp là bắt buộc'],
    index: true
  },
  carrierName: {
    type: String,
    required: [true, 'Tên nhà cung cấp là bắt buộc']
  },
  carrierLogo: {
    type: String,
    default: null
  },
  mode: {
    type: String,
    required: [true, 'Phương thức vận chuyển là bắt buộc'],
    enum: {
      values: ['flight', 'train', 'bus', 'ferry'],
      message: 'Phương thức vận chuyển không hợp lệ'
    },
    index: true
  },
  flightNumber: {
    type: String,
    sparse: true, // Only for flights
    index: true
  },
  trainNumber: {
    type: String,
    sparse: true, // Only for trains
    index: true
  },
  busNumber: {
    type: String,
    sparse: true, // Only for buses
    index: true
  },
  from: {
    type: String,
    required: [true, 'Điểm đi là bắt buộc'],
    index: true
  },
  fromCode: {
    type: String,
    required: [true, 'Mã điểm đi là bắt buộc'],
    index: true
  },
  to: {
    type: String,
    required: [true, 'Điểm đến là bắt buộc'],
    index: true
  },
  toCode: {
    type: String,
    required: [true, 'Mã điểm đến là bắt buộc'],
    index: true
  },
  departAt: {
    type: Date,
    required: [true, 'Thời gian khởi hành là bắt buộc'],
    index: true
  },
  arriveAt: {
    type: Date,
    required: [true, 'Thời gian đến là bắt buộc'],
    index: true
  },
  duration: {
    type: Number, // Duration in minutes
    required: true,
    min: 0
  },
  basePrice: {
    type: Number,
    required: [true, 'Giá cơ bản là bắt buộc'],
    min: 0
  },
  currency: {
    type: String,
    required: true,
    default: 'VND'
  },
  stops: [stopSchema],
  classOptions: [classOptionSchema],
  status: {
    type: String,
    enum: ['scheduled', 'boarding', 'departed', 'arrived', 'cancelled', 'delayed'],
    default: 'scheduled',
    index: true
  },
  gate: String,
  terminal: String,
  aircraft: String, // For flights
  trainType: String, // For trains
  busType: String, // For buses
  isActive: {
    type: Boolean,
    default: true,
    index: true
  },
  allowBooking: {
    type: Boolean,
    default: true
  },
  bookingDeadline: {
    type: Date,
    default: function() {
      // Default booking deadline is 1 hour before departure
      return new Date(this.departAt - 60 * 60 * 1000);
    }
  },
  cancellationDeadline: {
    type: Date,
    default: function() {
      // Default cancellation deadline is 2 hours before departure
      return new Date(this.departAt - 2 * 60 * 60 * 1000);
    }
  },
  metadata: {
    type: Map,
    of: mongoose.Schema.Types.Mixed,
    default: new Map()
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Compound indexes for efficient queries
tripSchema.index({ fromCode: 1, toCode: 1, departAt: 1 });
tripSchema.index({ mode: 1, departAt: 1 });
tripSchema.index({ carrierId: 1, departAt: 1 });
tripSchema.index({ status: 1, isActive: 1 });
tripSchema.index({ departAt: 1, status: 1 });

// Virtual for route (from-to combination)
tripSchema.virtual('route').get(function() {
  return `${this.fromCode}-${this.toCode}`;
});

// Virtual for available seats across all classes
tripSchema.virtual('totalAvailableSeats').get(function() {
  return this.classOptions.reduce((total, classOption) => total + classOption.availableSeats, 0);
});

// Virtual for total seats across all classes
tripSchema.virtual('totalSeats').get(function() {
  return this.classOptions.reduce((total, classOption) => total + classOption.totalSeats, 0);
});

// Virtual for lowest price
tripSchema.virtual('lowestPrice').get(function() {
  if (!this.classOptions || this.classOptions.length === 0) return this.basePrice;
  return Math.min(...this.classOptions.map(option => option.price));
});

// Virtual for booking availability
tripSchema.virtual('canBook').get(function() {
  const now = new Date();
  return this.isActive && 
         this.allowBooking && 
         this.status === 'scheduled' && 
         this.bookingDeadline > now &&
         this.totalAvailableSeats > 0;
});

// Pre-save middleware to calculate duration
tripSchema.pre('save', function(next) {
  if (this.isModified('departAt') || this.isModified('arriveAt')) {
    this.duration = Math.round((this.arriveAt - this.departAt) / (1000 * 60)); // Duration in minutes
  }
  next();
});

// Static method to search trips
tripSchema.statics.searchTrips = function(searchParams) {
  const {
    from,
    to,
    departDate,
    mode,
    minPrice,
    maxPrice,
    carrierId,
    page = 1,
    limit = 20,
    sortBy = 'departAt',
    sortOrder = 'asc'
  } = searchParams;

  const query = {
    isActive: true,
    status: 'scheduled',
    allowBooking: true
  };

  if (from) query.fromCode = from;
  if (to) query.toCode = to;
  if (mode) query.mode = mode;
  if (carrierId) query.carrierId = carrierId;

  // Date range for departure
  if (departDate) {
    const startOfDay = new Date(departDate);
    startOfDay.setHours(0, 0, 0, 0);
    const endOfDay = new Date(departDate);
    endOfDay.setHours(23, 59, 59, 999);
    
    query.departAt = {
      $gte: startOfDay,
      $lte: endOfDay
    };
  }

  // Price range
  if (minPrice || maxPrice) {
    const priceQuery = {};
    if (minPrice) priceQuery.$gte = minPrice;
    if (maxPrice) priceQuery.$lte = maxPrice;
    
    // Use aggregation to filter by price range across class options
    return this.aggregate([
      { $match: query },
      { $match: { 'classOptions.price': priceQuery } },
      { $sort: { [sortBy]: sortOrder === 'desc' ? -1 : 1 } },
      { $skip: (page - 1) * limit },
      { $limit: limit }
    ]);
  }

  return this.find(query)
    .sort({ [sortBy]: sortOrder === 'desc' ? -1 : 1 })
    .skip((page - 1) * limit)
    .limit(limit);
};

// Static method to get popular routes
tripSchema.statics.getPopularRoutes = function(limit = 10) {
  return this.aggregate([
    { $match: { isActive: true, status: 'scheduled' } },
    { 
      $group: {
        _id: { from: '$fromCode', to: '$toCode', fromName: '$from', toName: '$to' },
        count: { $sum: 1 },
        minPrice: { $min: '$basePrice' },
        avgPrice: { $avg: '$basePrice' }
      }
    },
    { $sort: { count: -1 } },
    { $limit: limit },
    {
      $project: {
        _id: 0,
        route: '$_id',
        tripCount: '$count',
        minPrice: '$minPrice',
        avgPrice: '$avgPrice'
      }
    }
  ]);
};

// Instance method to check seat availability
tripSchema.methods.checkSeatAvailability = function(classId, requestedSeats = 1) {
  const classOption = this.classOptions.find(option => option.id === classId);
  if (!classOption) return false;
  return classOption.availableSeats >= requestedSeats;
};

// Instance method to reserve seats
tripSchema.methods.reserveSeats = function(classId, seatCount) {
  const classOption = this.classOptions.find(option => option.id === classId);
  if (!classOption || classOption.availableSeats < seatCount) {
    throw new Error('Không đủ ghế trống');
  }
  classOption.availableSeats -= seatCount;
};

// Instance method to release seats
tripSchema.methods.releaseSeats = function(classId, seatCount) {
  const classOption = this.classOptions.find(option => option.id === classId);
  if (!classOption) return;
  classOption.availableSeats = Math.min(
    classOption.availableSeats + seatCount,
    classOption.totalSeats
  );
};

module.exports = mongoose.model('Trip', tripSchema);
